README.md
# Twitter_Bot

These are a few bots I scripted for the glug auditions.
All the bots are scripted in tweepy and python unless mentioned.

**1. glug_twitter_bot.py**
This bot checks the replies to all the tweets in which it is mentioned for a specific text, in this case, 'bot' and replies to that reply.

**2. search_for_#hobbist.py**
This is a multithreaded bot(only 2 threads for now) which searches for a tweet for a particular text or hashtag, #hobbist in this case, and replies to the tweet, 'I too support #hobbist' in this case.

**3. selenium_bot.py**
this a bot made with python and selenium which checks if the #2 bot is working.


history.txt
1363040314508775433

selenium_bot.py
from selenium import webdriver
from selenium.webdriver.common.keys import Keys
#from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import time

def account_info():
    with open('account_info.txt', 'r') as f:
        info = f.read().split()
        email = info[0]
        password = info[1]
    return email, password

email, password = account_info()

tweet = "This is a tweet"

PATH = "/home/diptangshudey/chromedriver_linux64/chromedriver"
driver = webdriver.Chrome(PATH)

driver.get("http://twitter.com/login")

email_xpath = '/html/body/div/div/div/div[2]/main/div/div/div[2]/form/div/div[1]/label/div/div[2]/div/input'
password_xpath = '/html/body/div/div/div/div[2]/main/div/div/div[2]/form/div/div[2]/label/div/div[2]/div/input'
login_xpath = '/html/body/div/div/div/div[2]/main/div/div/div[2]/form/div/div[3]/div/div'
profile_xpath = '/html/body/div/div/div/div[2]/header/div/div/div/div[1]/div[2]/nav/a[7]'
tweet_text = '/html/body/div/div/div/div[2]/main/div/div/div/div[1]/div/div[2]/div/div[2]/div[1]/div/div/div/div[2]/div[1]/div/div/div/div/div/div/div/div/div/div[1]/div/div/div/div[2]/div/div/div/div'
tweet_button = '/html/body/div/div/div/div[2]/main/div/div/div/div/div/div[2]/div/div[2]/div[1]/div/div/div/div[2]/div[4]/div/div/div[2]/div[3]'
tweet_1 = '/html/body/div/div/div/div[2]/main/div/div/div/div[1]/div/div[2]/div/div/div[2]/section/div/div/div[1]/div/div/article/div/div/div/div[2]'
tweet_page = '/html/body/div/div/div/div[2]/main/div/div/div/div/div/div[2]/div/section'
time.sleep(2)

driver.find_element_by_xpath(email_xpath).send_keys(email)
time.sleep(0.5)
driver.find_element_by_xpath(password_xpath).send_keys(password)
time.sleep(0.5)
driver.find_element_by_xpath(login_xpath).click()
time.sleep(5)

driver.find_element_by_xpath(tweet_text).send_keys('#hobbist Selenium test 4')
time.sleep(1)
driver.find_element_by_xpath(tweet_button).click()

time.sleep(2)

driver.find_element_by_xpath(profile_xpath).click()

time.sleep(1)
driver.find_element_by_xpath(tweet_1).click()

time.sleep(30)

driver.refresh()

try:
    if 'reply 1' in driver.find_element_by_xpath(tweet_page).text and '@DiptangshuDey' in driver.find_element_by_xpath(tweet_page).text:
        print('It works')
    else:
        print("It didn't work")
finally:
    driver.close()


glug_twitter_bot.py
import tweepy
import time

CONSUMER_KEY = ''
CONSUMER_SECRET = ''
ACCESS_KEY = ''
ACCESS_SECRET = ''
auth = tweepy.OAuthHandler(CONSUMER_KEY, CONSUMER_SECRET)
auth.set_access_token(ACCESS_KEY, ACCESS_SECRET)
api = tweepy.API(auth)

FILE_NAME = 'history.txt'
def retrieve_last_seen_id(file_name):
    f_read = open(file_name, 'r')
    last_seen_id = int(f_read.read().strip())
    f_read.close()
    return last_seen_id

def store_last_seen_id(last_seen_id, file_name):
    f_write = open(file_name, 'w')
    f_write.write(str(last_seen_id))
    f_write.close()
    return 

def reply_to_comment():
    last_seen_id = retrieve_last_seen_id(FILE_NAME)
    mention = api.mentions_timeline( tweet_mode = 'extended')

    for m in reversed(mention):
        tweet_id = m.id
        last_seen_id = retrieve_last_seen_id(FILE_NAME)
        
        print(m.full_text)
        for tweet in tweepy.Cursor(api.search, q='to:' + m.user.screen_name,since_id = last_seen_id, reselt_type = 'recent', timeout=999999).items(1000):
            try:
                if not hasattr(tweet, 'in_reply_to_status_id_str'):
                    continue
                if (tweet.in_reply_to_status_id_str == str(tweet_id)) and ('bot' in tweet.text.lower()) :
                    last_seen_id = tweet.id
                    store_last_seen_id(last_seen_id, FILE_NAME)
                    print(tweet.text)
                    print(tweet.id)
                    api.update_status('@' + tweet.user.screen_name + ' Yes I am a bot...', tweet.id)
                    time.sleep(15)
            except tweepy.RateLimitError as e:
                print("Twitter api rate is reached: {}".format(e))
                time.sleep(60)
                continue
            except tweepy.TweepError as e:
                print("Tweepy error reached: {}".format(e))
                break
            except StopIteration:
                print('bruh')
                break
            except Exception as e:
                print("Failed while fetching replies: {}".format(e))
                break

while True:
    reply_to_comment()
    time.sleep(60)


search_for_#hobbist.py
import tweepy
import time
import json
from threading import *

with open('twitter_credentials.txt') as f:
    data = f.read()

js = json.loads(data)
print('this my bot')

CONSUMER_KEY = js["CONSUMER_KEY"]
CONSUMER_SECRET = js["CONSUMER_SECRET"]
ACCESS_KEY = js["ACCESS_KEY"]
ACCESS_SECRET = js["ACCESS_SECRET"]

FILE_NAME = 'last_seen_id.txt'

auth = tweepy.OAuthHandler(CONSUMER_KEY, CONSUMER_SECRET)
auth.set_access_token(ACCESS_KEY, ACCESS_SECRET)
api = tweepy.API(auth)

def retrieve_last_seen_id(file_name):
    f_read = open(file_name, 'r')
    last_seen_id = int(f_read.read().strip())
    f_read.close()
    return last_seen_id

def store_last_seen_id(last_seen_id, file_name):
    f_write = open(file_name, 'w')
    f_write.write(str(last_seen_id))
    f_write.close()
    return 

ongoing = 123


class search(Thread):
    def run(self):
        print("started 1")
        global ongoing
        while True:
            FILE_NAME = 'last_seen_id.txt'

            f_read = open(FILE_NAME, 'r')
            last_seen_id = int(f_read.read().strip())
            f_read.close()
            
            for tweet in reversed(api.search( q='#hobbist', since_id = last_seen_id, result_type = 'recent',  timeout = 999999)):
                if ongoing != tweet.id:
                    try:
                        if tweet.user.screen_name != 'DiptangshuDey':
                            ongoing = tweet.id                       
                            f_write = open(FILE_NAME, 'w')
                            f_write.write(str(tweet.id))
                            f_write.close()

                            api.update_status('@' + tweet.user.screen_name + ' I support #hobbist from 1', tweet.id)

                            print(tweet.text)
                            print(tweet.id)

                            time.sleep(15)

                    except tweepy.RateLimitError as e:
                        print("Twitter api rate is reached: {}".format(e))
                        time.sleep(60)
                        continue
                    except tweepy.TweepError as e:
                        print("Tweepy error reached: {}".format(e))
                        break
                    except StopIteration:
                        print('bruh')
                        break
                    except Exception as e:
                        print("Failed while fetching replies: {}".format(e))
                        break
                else:
                    continue

class search2(Thread):
    def run(self):
        print("started 2")
        global ongoing
        while True:
            FILE_NAME = 'last_seen_id.txt'

            f_read = open(FILE_NAME, 'r')
            last_seen_id = int(f_read.read().strip())
            f_read.close()

            for tweet in reversed(api.search( q='#hobbist', since_id = last_seen_id, result_type = 'recent',  timeout = 999999)):
                if ongoing != tweet.id:
                    try:
                        if tweet.user.screen_name != 'DiptangshuDey':
                            ongoing = tweet.id
                            
                            f_write = open(FILE_NAME, 'w')
                            f_write.write(str(tweet.id))
                            f_write.close()

                            api.update_status('@' + tweet.user.screen_name + ' I support #hobbist from 2', tweet.id)

                            print(tweet.text)
                            print(tweet.id)

                            time.sleep(15)

                    except tweepy.RateLimitError as e:
                        print("Twitter api rate is reached: {}".format(e))
                        time.sleep(60 * 15)
                        continue
                    except tweepy.TweepError as e:
                        print("Tweepy error reached: {}".format(e))
                        time.sleep(60 * 15)
                        break
                    except StopIteration:
                        print('bruh')
                        break
                    except Exception as e:
                        print("Failed while fetching replies: {}".format(e))
                        break
                else:
                    continue


t1 = search()
t2 = search2()

t1.start()
time.sleep(5)
t2.start()